engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala

```diff
@@ -0,0 +1,32 @@
+package org.enso.compiler.core.ir
+
+import org.enso.compiler.core.IR
+
+trait LazyDiagnosticStorage { self: IR =>
+
+  private[this] var _diagnostics: DiagnosticStorage = _
+
+  protected def diagnostics_=(diagnostics: DiagnosticStorage): Unit = {
+    assert(_diagnostics eq null)
+    _diagnostics = diagnostics
+  }
+
+  override def diagnostics: DiagnosticStorage = {
+    _diagnostics
+  }
+
+  def diagnosticsList: List[Diagnostic] = {
+    if (_diagnostics eq null) Nil else _diagnostics.toList
+  }
+
+  def diagnosticsCopy: DiagnosticStorage = {
+    if (_diagnostics eq null) _diagnostics else _diagnostics.copy
+  }
+
+  override def getDiagnostics: DiagnosticStorage = {
+    if (_diagnostics eq null) {
+      _diagnostics = new DiagnosticStorage()
```

Comment:
This is not ready for multi threaded access. Are we OK with that?

Reply:
The entire compilation logic is not thread-safe. Compilation jobs ensure that the process is single-threaded

Reply:
Why are we then using immutable `IR` and not modifying `IR` in place? Ó the wonders of _like to be_ functional programming!